tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+			return errors.Errorf("invalid chunk format version %d", v)
+		}
+
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+8))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last segment. We set it to current time so that it will be eventually deleted.
+			f, err := os.OpenFile(files[i], os.O_WRONLY|os.O_CREATE, 0666)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			binary.BigEndian.PutUint64(w.buf[:], uint64(time.Now().Unix()))
+			if _, err := f.WriteAt(w.buf[:8], HeaderMaxtOffset); err != nil {
+				return err
+			}
+		}
+
+		w.size += int64(b.Len())
+	}
+
+	return nil
+}
+
+func sequenceFilesMap(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+func closeAllFromMap(cs map[int]io.Closer) error {
+	var merr tsdb_errors.MultiError
+
+	for _, c := range cs {
+		merr.Add(c.Close())
+	}
+	return merr.Err()
+}
+
+// IterateAllChunks iterates on all the chunks in it's byte slices in the order of the segment file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (w *HeadReadWriter) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	// Iterate files in ascending order.
+	seqs := make([]int, 0, len(w.bs))
+	for seg := range w.bs {
+		seqs = append(seqs, seg)
+	}
+	sort.Ints(seqs)
+	for _, seq := range seqs {
+		bs := w.bs[seq]
+		sliceLen := bs.Len()
+		idx := HeadSegmentHeaderSize
+		for idx < sliceLen {
+			if sliceLen-idx < 30 {
+				return errors.Errorf("segment doesn't include enough bytes to read the chunk header - required:%v, available:%v", idx+25, sliceLen)
+			}
+
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
+			idx += 8
+
+			mint := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
+				return err
+			}
+
+			idx++ // Skip encoding.
+			// Skip the data.
+			dataLen, n := binary.Uvarint(bs.Range(idx, idx+MaxChunkLengthFieldSize))
+			idx += n + int(dataLen)
+		}
+	}
+
+	return nil
+}
+func (w *HeadReadWriter) Close() error {
+	w.bsMtx.Lock()
+	defer w.bsMtx.Unlock()
+
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	if err := w.dirFile.Close(); err != nil {
+		return err
+	}
+
+	return closeAllFromMap(w.cs)
+}
+
+// WriteChunk writes chunk in the following format:
+// | Series Ref <8B> | MinT <8B> | MaxT <8B> | Chunk Encoding <1B> | Chunk Data Length <varint> | Chunk Data |
```

Comment:
I'd prefer we keep this in the docs, rather than duplicating it in the comments and one of them going stale.